design/010-sync-to-custom-resource.md

```diff
@@ -0,0 +1,113 @@
+```yaml
+---
+title: Sync to Custom Resources
+version: v1alpha1
+authors: Gustavo Carvalho
+creation-date: 2024-05-03
+status: draft
+---
+```
+
+# Sync to Custom Resources Design
+
+## Table of Contents
+
+<!-- toc -->
+// autogen please
+<!-- /toc -->
+
+## Summary
+
+This design document describes how `ExternalSecrets` can leverage templates to generate non-kubernetes `Secret` resource as the target. This allows to push Sensitive information to specific CRs, `ConfiMaps`, etc.
+
+## Motivation
+
+Currently, several "semi-sensitive" information needs to be provisioned directly into CRs and ConfigMaps (such as OIDC Client IDs). While these information are not strictly speaking a secret, several regulated environments must treat them as such, causing several operational overhead to deal with this information - specially on a GitOps setup.
+
+## Proposal
+
+To simplify the workflow and enhance user experience, the proposal is to integrate a functionality to template the whole manifest. this would be additional logic to the existing `target: Manifest` directly into the `templateFrom.[].target` resource. This will allow users to specify a template to render any type of manifest, instead of the original Secret object.
+
+Problems with this proposal is that the whole reconciliation logic reads from a secret - this would need to be updated if templates using Manifests are set. In that case, we should query for that specific resource as specified on the target.
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+spec:
+  target:
+    name: target-custom-resource
+    manifest: # Information as to find it afterwards during reconcile cycle
+      APIVersion: my.custom.api/v1alpha1
+      Kind: CustomResource
+    template:
+      templateFrom:
+      - target: Spec #Additional target at root level (instead of data, metadata and annotations level).
+      # Other option would be to allow a gjson path such as target:'.spec' where . is the indicator of this type of expression for backwards compatibility.
+        literal: |
+            customSpecField1: {{ .fromSecretStore }}
+            field2:
+              couldBeNested:
+              - {{ .fromSecretStore }}
+          ## Name: is obtained from spec.target.name
+---
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+spec:
+  target:
+    name: target-configmap
+    manifest: # Information as to find it afterwards during reconcile cycle
+      APIVersion: v1
+      Kind: ConfigMap
```

Comment:
Allowing this opens the door for a lot of use cases in platform engineering beyond GitOps, but also the ability to use ESO to configure the CNCF Backstage configMap when running on Kubernetes.